File: tpu_sync/api/torch/reshard_client.py
# Copyright 2026 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""C++-owned reshard client.

Drop-in for the reshard command surface of
tpu_raiden.rpc.raiden_controller.RaidenControllerClientFacade: identical
method signatures, return shapes, and exception text — the encoding,
presence rules, and transport live in C++ (kv_cache/reshard/reshard_client).
This shim only normalizes Python object shapes (duck-typed RaidenId and
span entries, pb2 variable protos) into plain values for the binding, and
parses get_metadata payloads back into pb2 objects so callers see the
facade's return type.

The legacy relay surface (register_transfer_schedule and dense broadcast) is
deliberately not provided here and remains on the Python facade.
"""

import typing
from typing import Any, Optional

from tpu_sync.api.torch import torch_tpu_common_loader

torch_tpu_common_loader.load_torch_tpu_common()

# pylint: disable=g-import-not-at-top
from tpu_sync.api.torch import torch_abi

_impl = torch_abi.load_extension(
    "tpu_sync.frameworks.torch",
    "_tpu_raiden_torch",
)
# pylint: enable=g-import-not-at-top

# Capability probe for the vLLM connector: True iff this client (and the
# native extension behind it) accepts the per-tag dst_skip_bytes clip on
# coordinate_transfer. Absent on older shims, so consumers must getattr.
SUPPORTS_DST_SKIP_BYTES = bool(
    getattr(_impl, "reshard_client_supports_dst_skip_bytes", False))
# True iff get_request_block_status (read-only registry lifecycle probe)
# is available; absent on older shims, so consumers must getattr.
SUPPORTS_REQUEST_BLOCK_STATUS = bool(
    getattr(_impl, "reshard_client_supports_request_block_status", False)
)

# GetRequestBlockStatusResponse.Status values returned by
# get_request_block_status.
REQUEST_BLOCK_STATUS_UNKNOWN = 1
REQUEST_BLOCK_STATUS_REGISTERED = 2
REQUEST_BLOCK_STATUS_CLAIMED = 3
REQUEST_BLOCK_STATUS_CANCELLED = 4


def _unit_tuple(unit: Any) -> tuple:
  """Duck-typed: accepts the rpc dataclass, the bound RaidenId, or alikes."""
  return (
      str(getattr(unit, "job_name", "")),
      str(getattr(unit, "job_replica_id", "")),
      str(getattr(unit, "data_name", "")),
      int(getattr(unit, "data_replica_idx", 0)),
  )


def _dst_unit_ordinal(span: Any) -> int:
  """-1 = every destination (absent on the wire); >= 0 = one destination."""
  value = getattr(span, "dst_unit_ordinal", None)
  return -1 if value is None else int(value)


def _span_entry_dict(entry: Any) -> dict:
  return {
      "tag": str(entry.tag),
      "block_ids": [int(block_id) for block_id in entry.block_ids],
      "declared_bytes": int(entry.declared_bytes),
      "dst_space_version": int(getattr(entry, "dst_space_version", 0)),
      "spans": [
          (
              int(span.src_block_ordinal),
              int(span.src_offset_bytes),
              int(span.dst_block_index),
              int(span.dst_offset_bytes),
              int(span.size_bytes),
              int(span.src_stride_bytes),
              int(span.dst_stride_bytes),
              int(span.count),
              _dst_unit_ordinal(span),
          )
          for span in entry.spans
      ],
  }


def _pool_dict(pool: Any) -> dict:
  """Mirrors _coerce_pool_spec_proto's mapping/dataclass tolerance."""
  if isinstance(pool, typing.Mapping):
    get = pool.get
  else:
    get = lambda name, default=None: getattr(pool, name, default)  # noqa: E731
  regions = []
  for region in get("regions", ()) or ():
    if isinstance(region, typing.Mapping):
      rget = region.get
    else:
      rget = lambda name, default=None, r=region: getattr(r, name, default)  # noqa: E731
    regions.append({
        "name": str(rget("name", "")),
        "offset_bytes": int(rget("offset_bytes", 0)),
        "stride_bytes": int(rget("stride_bytes", 0)),
        "unit_bytes": int(rget("unit_bytes", 0)),
        "num_units": int(rget("num_units", 0)),
        "units_per_stride": int(rget("units_per_stride", 1)),
    })
  return {
      "tag": str(get("tag", "")),
      "storage_index": int(get("storage_index", 0)),
      "base_offset_bytes": int(get("base_offset_bytes", 0)),
      "block_stride_bytes": int(get("block_stride_bytes", 0)),
      "num_blocks": int(get("num_blocks", 0)),
      "dtype_tag": str(get("dtype_tag", "")),
      "regions": regions,
  }


class ReshardClient:
  """Facade-compatible reshard client backed by the C++ implementation."""

  def __init__(self, controller_address: str, **_unused: Any):
    # name_resolver/proto_module kwargs of the Python facade are accepted
    # and ignored: name resolution is the transport's job in C++, and the
    # proto modules are fixed.
    self._address = controller_address
    self._impl = _impl.ReshardClient(controller_address=controller_address)

  def register_work_unit(
      self,
      unit: Any,
      shards: list,
      control_plane_rpc_address: Optional[str] = None,
      mesh_shape: Optional[typing.Sequence[int]] = None,
      layout: Optional[typing.Sequence[int]] = None,
      global_shape: Optional[typing.Sequence[int]] = None,
      itemsize: Optional[int] = None,
      pool_manifest: Optional[typing.Sequence[Any]] = None,
      layout_fingerprint: Optional[str] = None,
      page_tokens: Optional[int] = None,
      transfer_parallelism: Optional[int] = None,
      transfer_rank: Optional[int] = None,
      variables: Optional[typing.Sequence[Any]] = None,
  ) -> None:
    self._impl.register_work_unit(
        _unit_tuple(unit),
        [str(s) for s in shards],
        control_plane_rpc_address if control_plane_rpc_address else "",
        [int(v) for v in mesh_shape] if mesh_shape else [],
        [int(v) for v in layout] if layout else [],
        [int(v) for v in global_shape] if global_shape else [],
        int(itemsize) if itemsize else 0,
        pool_manifest is not None,
        [_pool_dict(p) for p in (pool_manifest or ())],
        layout_fingerprint,
        int(page_tokens) if page_tokens is not None else None,
        int(transfer_parallelism)
        if transfer_parallelism is not None else None,
        int(transfer_rank) if transfer_rank is not None else None,
        variables is not None,
        [v.SerializeToString() for v in (variables or ())],
    )

  def register_request_blocks(
      self,
      req_id: str,
      uuid: int,
      unit: Any,
      block_ids: typing.Sequence[int],
      pool_spans: typing.Sequence[Any] = (),
  ) -> None:
    self._impl.register_request_blocks(
        str(req_id), int(uuid), _unit_tuple(unit),
        [int(b) for b in block_ids],
        [_span_entry_dict(entry) for entry in pool_spans])

  def release_request_blocks(self, req_id: str, uuid: int) -> None:
    self._impl.release_request_blocks(str(req_id), int(uuid))

  def complete_request_blocks(self, req_id: str, uuid: int,
                              unit: Any) -> None:
    self._impl.complete_request_blocks(str(req_id), int(uuid),
                                       _unit_tuple(unit))

  def cancel_request_blocks_if_unclaimed(self, req_id: str,
                                         uuid: int) -> bool:
    return bool(
        self._impl.cancel_request_blocks_if_unclaimed(str(req_id), int(uuid)))

  def get_request_block_status(
      self, keys: typing.Sequence[tuple[str, int]]
  ) -> list[int]:
    """Read-only registry lifecycle probe: one REQUEST_BLOCK_STATUS_* value

    per (req_id, uuid) key, in order.
    """
    return [
        int(status)
        for status in self._impl.get_request_block_status(
            [(str(req_id), int(uuid)) for req_id, uuid in keys]
        )
    ]

  def coordinate_transfer(
      self,
      src_units: list,
      dst_units: list,
      req_id: Optional[str] = None,
      use_block_chunks: bool = False,
      is_sender: bool = True,
      expected_block_count: int = 0,
      uuid: int = 0,
      dst_controller_address: Optional[str] = None,
      src_controller_address: Optional[str] = None,
      shard_push_schedules: Optional[dict] = None,
      dst_mem_type: Any = None,
      dst_device_block_ids: Optional[typing.Sequence[int]] = None,
      num_tokens: Optional[int] = None,
      transfer_pool_tags: Optional[typing.Sequence[str]] = None,
      dst_block_counts: Optional[typing.Sequence[int]] = None,
      dst_skip_bytes: Optional[typing.Sequence[int]] = None,
  ) -> bool:
    # num_tokens accepted for caller compatibility, retired from the wire
    # (facade parity); shard_push_schedules is a legacy-relay parameter the
    # pool path never sets.
    del num_tokens
    if shard_push_schedules:
      raise RuntimeError(
          "shard_push_schedules is a legacy relay parameter; the C++ "
          "reshard client serves the pool path only")
    if dst_mem_type is None:
      mem_type = 1  # RaidenMemoryType.DRAM, the facade default
    else:
      mem_type = int(dst_mem_type)
    return bool(
        self._impl.start_transfer(
            [_unit_tuple(u) for u in src_units],
            [_unit_tuple(u) for u in dst_units],
            req_id if req_id else "",
            bool(use_block_chunks),
            bool(is_sender),
            int(expected_block_count),
            int(uuid),
            dst_controller_address if dst_controller_address else "",
            src_controller_address if src_controller_address else "",
            mem_type,
            dst_device_block_ids is not None,
            [int(b) for b in (dst_device_block_ids or ())],
            transfer_pool_tags is not None,
            [str(t) for t in (transfer_pool_tags or ())],
            [int(c) for c in (dst_block_counts or ())]
            if dst_block_counts else [],
            [int(s) for s in (dst_skip_bytes or ())]
            if dst_skip_bytes and any(int(s) != 0 for s in dst_skip_bytes)
            else [],
        ))

  def start_transfer(self, *args, **kwargs) -> bool:
    """Alias for coordinate_transfer for backward compatibility."""
    return self.coordinate_transfer(*args, **kwargs)

  def get_transfer_status(self, req_id: str, uuid: int = 0) -> int:
    return int(self._impl.get_transfer_status(str(req_id), int(uuid)))

  def get_metadata(self) -> list:
    from tpu_sync.rpc import raiden_service_pb2  # pylint: disable=g-import-not-at-top

    out = []
    for payload in self._impl.get_metadata():
      metadata = raiden_service_pb2.RegisterWorkUnitRequest()
      metadata.ParseFromString(payload)
      out.append(metadata)
    return out

  def shutdown(self) -> bool:
    return bool(self._impl.shutdown())

File: tpu_sync/kv_cache/reshard/reshard_client.cc
// Copyright 2026 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "tpu_sync/kv_cache/reshard/reshard_client.h"

#include <cstdint>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/time/time.h"
#include "tpu_sync/common/raiden_id.h"
#include "tpu_sync/kv_cache/reshard/framed_rpc.h"
#include "tpu_sync/rpc/controller_service.pb.h"
#include "tpu_sync/rpc/raiden_service.pb.h"

namespace tpu_raiden {
namespace kv_cache {
namespace reshard {

namespace {

// Facade parity: connect_socket(address, timeout=300.0) per call.
constexpr absl::Duration kCallTimeout = absl::Seconds(300);

tpu_sync::rpc::RaidenIdProto RaidenIdProtoOf(const RaidenId& unit) {
  tpu_sync::rpc::RaidenIdProto proto;
  proto.set_job_name(unit.job_name);
  proto.set_job_replica_id(unit.job_replica_id);
  proto.set_data_name(unit.data_name);
  proto.set_data_replica_idx(unit.data_replica_idx);
  return proto;
}

}  // namespace

ReshardClient::ReshardClient(std::string address, FramedTransport* transport)
    : address_(std::move(address)) {
  if (transport == nullptr) {
    owned_transport_ = std::make_unique<SocketFramedTransport>();
    transport_ = owned_transport_.get();
  } else {
    transport_ = transport;
  }
}

tpu_sync::rpc::ControlRequest ReshardClient::BuildRegisterWorkUnit(
    const RegisterWorkUnitArgs& args) {
  tpu_sync::rpc::ControlRequest req;
  req.set_command(tpu_sync::rpc::ControlRequest::COMMAND_REGISTER_WORK_UNIT);
  tpu_sync::rpc::RegisterWorkUnitRequest* reg =
      req.mutable_register_work_unit_request();
  *reg->mutable_unit() = RaidenIdProtoOf(args.unit);
  for (const std::string& shard : args.shards) reg->add_shards(shard);
  // Python assigns "" when None; proto3 implicit presence drops it either
  // way, so unconditional set_ matches the wire exactly.
  reg->set_control_plane_rpc_address(args.control_plane_rpc_address);
  // `if mesh_shape:` — present-but-empty is absent (same for layout and
  // global_shape); `if itemsize:` — zero is absent.
  for (int64_t v : args.mesh_shape) reg->add_mesh_shape(v);
  for (int32_t v : args.layout) reg->add_layout(v);
  for (int64_t v : args.global_shape) reg->add_global_shape(v);
  if (args.itemsize != 0) reg->set_itemsize(args.itemsize);
  if (args.has_pool_manifest) {
    for (const ClientPoolSpec& pool : args.pool_manifest) {
      tpu_sync::rpc::PoolSpecProto* pool_proto = reg->add_pools();
      pool_proto->set_tag(pool.tag);
      pool_proto->set_storage_index(pool.storage_index);
      pool_proto->set_base_offset_bytes(pool.base_offset_bytes);
      pool_proto->set_block_stride_bytes(pool.block_stride_bytes);
      pool_proto->set_num_blocks(pool.num_blocks);
      pool_proto->set_dtype_tag(pool.dtype_tag);
      for (const ClientPoolRegion& region : pool.regions) {
        tpu_sync::rpc::RegionSpecProto* region_proto =
            pool_proto->add_regions();
        region_proto->set_name(region.name);
        region_proto->set_offset_bytes(region.offset_bytes);
        region_proto->set_stride_bytes(region.stride_bytes);
        region_proto->set_unit_bytes(region.unit_bytes);
        region_proto->set_num_units(region.num_units);
        region_proto->set_units_per_stride(region.units_per_stride);
      }
    }
  }
  if (args.layout_fingerprint.has_value()) {
    reg->set_layout_fingerprint(*args.layout_fingerprint);
  }
  if (args.page_tokens.has_value()) reg->set_page_tokens(*args.page_tokens);
  if (args.transfer_parallelism.has_value()) {
    reg->set_transfer_parallelism(*args.transfer_parallelism);
  }
  if (args.transfer_rank.has_value()) {
    reg->set_transfer_rank(*args.transfer_rank);
  }
  if (args.has_variables) {
    for (const std::string& payload : args.variables) {
      // Byte-exact pass-through of the caller's VariableMetadataProto.
      reg->add_variables()->ParseFromString(payload);
    }
  }
  return req;
}

tpu_sync::rpc::ControllerRequest ReshardClient::BuildRegisterRequestBlocks(
    const std::string& req_id, int64_t uuid, const RaidenId& unit,
    const std::vector<int64_t>& block_ids,
    const std::vector<ClientPoolSpans>& pool_spans) {
  tpu_sync::rpc::ControllerRequest req;
  req.set_command(
      tpu_sync::rpc::ControllerRequest::COMMAND_REGISTER_REQUEST_BLOCKS);
  tpu_sync::rpc::RegisterRequestBlocksRequest* block_req =
      req.mutable_register_request_blocks_request();
  block_req->set_req_id(req_id);
  block_req->set_uuid(uuid);
  *block_req->mutable_unit() = RaidenIdProtoOf(unit);
  for (int64_t block_id : block_ids) block_req->add_block_ids(block_id);
  for (const ClientPoolSpans& entry : pool_spans) {
    auto* entry_proto = block_req->add_pool_spans();
    entry_proto->set_tag(entry.tag);
    for (int64_t block_id : entry.block_ids) {
      entry_proto->add_block_ids(block_id);
    }
    entry_proto->set_declared_bytes(entry.declared_bytes);
    entry_proto->set_dst_space_version(entry.dst_space_version);
    for (const ClientByteSpan& span : entry.spans) {
      auto* span_proto = entry_proto->add_spans();
      span_proto->set_src_block_ordinal(span.src_block_ordinal);
      span_proto->set_src_offset_bytes(span.src_offset_bytes);
      span_proto->set_dst_block_index(span.dst_block_index);
      span_proto->set_dst_offset_bytes(span.dst_offset_bytes);
      span_proto->set_size_bytes(span.size_bytes);
      span_proto->set_src_stride_bytes(span.src_stride_bytes);
      span_proto->set_dst_stride_bytes(span.dst_stride_bytes);
      span_proto->set_count(span.count);
      if (span.dst_unit_ordinal >= 0) {
        span_proto->set_dst_unit_ordinal(span.dst_unit_ordinal);
      }
    }
  }
  return req;
}

tpu_sync::rpc::ControllerRequest ReshardClient::BuildReleaseRequestBlocks(
    const std::string& req_id, int64_t uuid) {
  tpu_sync::rpc::ControllerRequest req;
  req.set_command(
      tpu_sync::rpc::ControllerRequest::COMMAND_RELEASE_REQUEST_BLOCKS);
  auto* release = req.mutable_release_request_blocks_request();
  release->set_req_id(req_id);
  release->set_uuid(uuid);
  release->set_force(true);  // facade: force=True always
  return req;
}

tpu_sync::rpc::ControllerRequest ReshardClient::BuildCompleteRequestBlocks(
    const std::string& req_id, int64_t uuid, const RaidenId& unit) {
  tpu_sync::rpc::ControllerRequest req;
  req.set_command(
      tpu_sync::rpc::ControllerRequest::COMMAND_COMPLETE_REQUEST_BLOCKS);
  auto* complete = req.mutable_complete_request_blocks_request();
  complete->set_req_id(req_id);
  complete->set_uuid(uuid);
  *complete->mutable_unit() = RaidenIdProtoOf(unit);
  return req;
}

tpu_sync::rpc::ControllerRequest
ReshardClient::BuildCancelRequestBlocksIfUnclaimed(const std::string& req_id,
                                                   int64_t uuid) {
  tpu_sync::rpc::ControllerRequest req;
  req.set_command(tpu_sync::rpc::ControllerRequest::
                      COMMAND_CANCEL_REQUEST_BLOCKS_IF_UNCLAIMED);
  auto* cancel = req.mutable_cancel_request_blocks_if_unclaimed_request();
  cancel->set_req_id(req_id);
  cancel->set_uuid(uuid);
  return req;
}

tpu_sync::rpc::ControllerRequest ReshardClient::BuildStartTransfer(
    const StartTransferArgs& args) {
  tpu_sync::rpc::ControllerRequest req;
  req.set_command(
      tpu_sync::rpc::ControllerRequest::COMMAND_COORDINATE_TRANSFER);
  tpu_sync::rpc::CoordinateTransferRequest* coord =
      req.mutable_coordinate_transfer_request();
  for (const RaidenId& unit : args.src_units) {
    *coord->add_src_units() = RaidenIdProtoOf(unit);
  }
  for (const RaidenId& unit : args.dst_units) {
    *coord->add_dst_units() = RaidenIdProtoOf(unit);
  }
  coord->set_use_block_chunks(args.use_block_chunks);
  coord->set_is_sender(args.is_sender);
  coord->set_expected_block_count(args.expected_block_count);
  coord->set_uuid(args.uuid);
  coord->set_req_id(args.req_id);
  coord->set_dst_controller_address(args.dst_controller_address);
  coord->set_src_controller_address(args.src_controller_address);
  coord->set_dst_mem_type(
      static_cast<tpu_sync::rpc::MemoryType>(args.dst_mem_type));
  if (args.has_dst_device_block_ids) {
    for (int64_t block : args.dst_device_block_ids) {
      coord->add_dst_device_block_ids(block);
    }
  }
  // num_tokens: accepted by the shim for caller compatibility, retired
  // from the wire (facade parity).
  if (args.has_transfer_pool_tags) {
    for (const std::string& tag : args.transfer_pool_tags) {
      coord->add_transfer_pool_tags(tag);
    }
  }
  // `if dst_block_counts:` — truthy; empty is absent.
  for (int64_t count : args.dst_block_counts) {
    coord->add_dst_block_counts(static_cast<int32_t>(count));
  }
  // `if dst_skip_bytes:` — truthy; empty (all-zero clip) is absent, which
  // keeps skip-free requests byte-identical to pre-clip encodings.
  for (int64_t skip : args.dst_skip_bytes) {
    coord->add_dst_skip_bytes(skip);
  }
  return req;
}

tpu_sync::rpc::ControllerRequest ReshardClient::BuildGetTransferStatus(
    const std::string& req_id, int64_t uuid) {
  tpu_sync::rpc::ControllerRequest req;
  req.set_command(
      tpu_sync::rpc::ControllerRequest::COMMAND_GET_TRANSFER_STATUS);
  auto* status = req.mutable_get_transfer_status_request();
  status->set_req_id(req_id);
  status->set_uuid(uuid);
  return req;
}

tpu_sync::rpc::ControllerRequest ReshardClient::BuildGetRequestBlockStatus(
    const std::vector<std::pair<std::string, int64_t>>& keys) {
  tpu_sync::rpc::ControllerRequest req;
  req.set_command(
      tpu_sync::rpc::ControllerRequest::COMMAND_GET_REQUEST_BLOCK_STATUS);
  auto* status = req.mutable_get_request_block_status_request();
  for (const auto& [req_id, uuid] : keys) {
    auto* key = status->add_keys();
    key->set_req_id(req_id);
    key->set_uuid(uuid);
  }
  return req;
}

tpu_sync::rpc::ControlRequest ReshardClient::BuildGetMetadata() {
  tpu_sync::rpc::ControlRequest req;
  req.set_command(tpu_sync::rpc::ControlRequest::COMMAND_GET_METADATA);
  return req;
}

tpu_sync::rpc::ControlRequest ReshardClient::BuildShutdown() {
  tpu_sync::rpc::ControlRequest req;
  req.set_command(tpu_sync::rpc::ControlRequest::COMMAND_SHUTDOWN);
  return req;
}

absl::StatusOr<tpu_sync::rpc::ControllerResponse> ReshardClient::CallController(
    const tpu_sync::rpc::ControllerRequest& request) {
  absl::StatusOr<std::string> body =
      transport_->Call(address_, request.SerializeAsString(), kCallTimeout);
  if (!body.ok()) return body.status();
  tpu_sync::rpc::ControllerResponse response;
  if (!response.ParseFromString(*body)) {
    return absl::InternalError("Failed to parse ControllerResponse");
  }
  if (!response.success()) {
    // Verbatim facade text: the connector's bounded retry substring-matches
    // the embedded server message.
    return absl::InternalError(absl::StrCat(
        "Remote Controller Server execution failed: ", response.message()));
  }
  return response;
}

absl::StatusOr<tpu_sync::rpc::ControlResponse> ReshardClient::CallRaiden(
    const tpu_sync::rpc::ControlRequest& request) {
  absl::StatusOr<std::string> body =
      transport_->Call(address_, request.SerializeAsString(), kCallTimeout);
  if (!body.ok()) return body.status();
  tpu_sync::rpc::ControlResponse response;
  if (!response.ParseFromString(*body)) {
    return absl::InternalError("Failed to parse ControlResponse");
  }
  if (!response.success()) {
    return absl::InternalError(absl::StrCat(
        "Remote Controller Server execution failed: ", response.message()));
  }
  return response;
}

absl::Status ReshardClient::RegisterWorkUnit(
    const RegisterWorkUnitArgs& args) {
  return CallRaiden(BuildRegisterWorkUnit(args)).status();
}

absl::Status ReshardClient::RegisterRequestBlocks(
    const std::string& req_id, int64_t uuid, const RaidenId& unit,
    const std::vector<int64_t>& block_ids,
    const std::vector<ClientPoolSpans>& pool_spans) {
  return CallController(BuildRegisterRequestBlocks(req_id, uuid, unit,
                                                   block_ids, pool_spans))
      .status();
}

absl::Status ReshardClient::ReleaseRequestBlocks(const std::string& req_id,
                                                 int64_t uuid) {
  return CallController(BuildReleaseRequestBlocks(req_id, uuid)).status();
}

absl::Status ReshardClient::CompleteRequestBlocks(const std::string& req_id,
                                                  int64_t uuid,
                                                  const RaidenId& unit) {
  return CallController(BuildCompleteRequestBlocks(req_id, uuid, unit))
      .status();
}

absl::StatusOr<bool> ReshardClient::CancelRequestBlocksIfUnclaimed(
    const std::string& req_id, int64_t uuid) {
  absl::StatusOr<tpu_sync::rpc::ControllerResponse> response =
      CallController(BuildCancelRequestBlocksIfUnclaimed(req_id, uuid));
  if (!response.ok()) return response.status();
  if (response->response_data() == "true") return true;
  if (response->response_data() == "false") return false;
  return absl::InternalError(absl::StrCat(
      "Remote Controller Server returned an invalid cancellation result: '",
      response->response_data(), "'"));
}

absl::StatusOr<bool> ReshardClient::StartTransfer(
    const StartTransferArgs& args) {
  absl::Status status = CallController(BuildStartTransfer(args)).status();
  if (!status.ok()) return status;
  return true;
}

absl::StatusOr<int32_t> ReshardClient::GetTransferStatus(
    const std::string& req_id, int64_t uuid) {
  absl::StatusOr<tpu_sync::rpc::ControllerResponse> response =
      CallController(BuildGetTransferStatus(req_id, uuid));
  if (!response.ok()) return response.status();
  return static_cast<int32_t>(
      response->get_transfer_status_response().status());
}

absl::StatusOr<std::vector<int32_t>> ReshardClient::GetRequestBlockStatus(
    const std::vector<std::pair<std::string, int64_t>>& keys) {
  absl::StatusOr<tpu_sync::rpc::ControllerResponse> response =
      CallController(BuildGetRequestBlockStatus(keys));
  if (!response.ok()) return response.status();
  const auto& body = response->get_request_block_status_response();
  if (static_cast<size_t>(body.statuses_size()) != keys.size()) {
    return absl::InternalError(
        absl::StrCat("Remote Controller Server returned ", body.statuses_size(),
                     " request block statuses for ", keys.size(), " keys"));
  }
  std::vector<int32_t> statuses;
  statuses.reserve(keys.size());
  for (int status : body.statuses()) statuses.push_back(status);
  return statuses;
}

absl::StatusOr<std::vector<std::string>> ReshardClient::GetMetadata() {
  absl::StatusOr<tpu_sync::rpc::ControlResponse> response =
      CallRaiden(BuildGetMetadata());
  if (!response.ok()) return response.status();
  std::vector<std::string> serialized;
  serialized.reserve(response->get_metadata_response().metadata_size());
  for (const auto& metadata : response->get_metadata_response().metadata()) {
    serialized.push_back(metadata.SerializeAsString());
  }
  return serialized;
}

absl::StatusOr<bool> ReshardClient::Shutdown() {
  absl::Status status = CallRaiden(BuildShutdown()).status();
  if (!status.ok()) return status;
  return true;
}

}  // namespace reshard
}  // namespace kv_cache
}  // namespace tpu_raiden
